torch/csrc/jit/passes/annotate_effects.cpp

```diff
@@ -0,0 +1,184 @@
+      if (!shouldAddFence(block)) {
+        // Bail out early if there's no mutable variables used inside
+        return curToken;
+      }
+
+      // Register the world token as a loop carried dependency
+      auto beginLoopToken = block->addInput()->setType(WorldType::get());
+      auto endLoopToken = visitBlock(block, beginLoopToken);
+      block->registerOutput(endLoopToken);
+
+      JIT_ASSERT(endLoopToken != beginLoopToken);
+
+      // Thread the world token through the loop node
+      node->addInput(curToken);
+      return node->addOutput()->setType(WorldType::get());
+    }
+
+    JIT_ASSERT(node->blocks().size() == 0);
+
+    if (shouldAddFence(node)) {
+      return addFenceForNode(node, curToken);
+    }
+
+    return curToken;
+  }
+
+  bool shouldAddFence(Node* node) {
+    // Check if this node uses a known mutable value
+    for (auto* input : node->inputs()) {
+      if (mutableValues_.count(input) != 0) {
+        return true;
+      }
+    }
+
+    // Otherwise, check if this node is the first mutation of a value
+    if (effectfulOps.count(node->kind()) != 0) {
+      const auto inputs = node->inputs();
+      Value* mut = inputs.at(0);
+      JIT_ASSERT(mut->type()->kind() == TypeKind::ListType); // TODO
+
+      mutableValues_.insert(mut);
+      return true;
+    }
+
+    // Check that any sub-blocks
+    for (auto block : node->blocks()) {
+      if (shouldAddFence(block)) {
+        return true;
+      }
+    }
+
+    return false;
+  }
+
+  bool shouldAddFence(Block* block) {
+    for (auto node : block->nodes()) {
+      if (shouldAddFence(node)) {
+        return true;
+      }
+    }
+    return false;
+  }
+
+
+  // Create a memory fence around a node, using the world token
+  // Input:
+  //   = aten::append(%list, %new_element)
+  //
+  // Output:
+  //  %t.1 : World, %list.2 : int[] = prim::MemoryFence(%curToken, %list)
+  //   = aten::append(%list.2, %new_element)
+  //  %t.2 : World, %list.3 : int[] = prim::MemoryFence(%t.1, %list2)
+  //
+  // Returns the new world token (%t.2) for subsequent fences to use.
+  Value* addFenceForNode(Node* node, Value* curToken) {
+    auto mut = node->inputs().at(0);
+
+    // Add a start fence
+    auto startFence =
+        node->owningGraph()->create(prim::MemoryFence, /*outputs=*/2);
+    startFence->addInput(curToken);
+    startFence->addInput(mut);
+    curToken = startFence->outputs()[0]->setType(WorldType::get());
+    mut = startFence->outputs()[1]->setType(ListType::ofInts());
+    startFence->insertBefore(node);
+
+    // modify the node to take in the start fence's output list
+    node->replaceInput(0, mut);
+
+    // Add an end fence
+    auto endFence =
+        node->owningGraph()->create(prim::MemoryFence, /*outputs=*/2);
+    endFence->addInput(curToken);
+    endFence->addInput(mut);
+    curToken = endFence->outputs()[0]->setType(WorldType::get());
+    mut = endFence->outputs()[1]->setType(ListType::ofInts());
+    endFence->insertAfter(node);
```

Comment:
Didn't you forget to replace all uses of the mutated thing with whatever the end fence returns?